lib/spack/spack/solver/asp.py

```diff
@@ -1843,22 +1846,34 @@ def generate_possible_compilers(self, specs):
                 if s.concrete:
                     continue
 
-                if not s.compiler or not s.compiler.concrete:
+                if not s.compiler:
                     continue
 
-                if strict and s.compiler not in cspecs:
-                    if not s.concrete:
-                        raise spack.concretize.UnavailableCompilerVersionError(s.compiler)
-                    # Allow unknown compilers to exist if the associated spec
-                    # is already built
-                else:
-                    compiler_cls = spack.compilers.class_for_compiler_name(s.compiler.name)
-                    compilers.append(
-                        compiler_cls(
-                            s.compiler, operating_system=None, target=None, paths=[None] * 4
+                # This is only for backward compatibility, where old spack
+                # would interpret %gcc@10.0 as %gcc@=10.0 when bootstrapping
+                # compilers.
+                compiler_version = s.compiler.versions.concrete_range_as_version
```

Comment:
I'd be ok with changing this.  But it should probably be in another PR and should go in the 0.20 changelog.

Reply:
This is now changed such that it `spack install pkg@1.2.3` and `spack install pkg %compiler@1.2.3` behave the same.

They fail if `1.2.3` does not match a known package / compiler version.

For defining the version on the fly, you need `@=`.